checkstyle/src/main.rs

```diff
@@ -0,0 +1,212 @@
+    let exclude_patterns = args.exclude;
+
+    let nonmatching_files = separate_regex_matching_files(
+        &regex_file_path,
+        &target_path,
+        include_patterns,
+        exclude_patterns,
+    );
+
+    for file in &nonmatching_files {
+        println!(
+            "File {} does not match regex: {}",
+            file.display(),
+            regex_file_path.display()
+        );
+    }
+
+    if !nonmatching_files.is_empty() {
+        process::exit(1);
+    }
+
+    Ok(())
+}
+
+fn read_file_content(file_path: &Path) -> Option<String> {
+    match fs::read_to_string(file_path) {
+        Ok(content) => Some(content),
+        Err(err) => {
+            if err.kind() == std::io::ErrorKind::InvalidData {
+                eprintln!(
+                    "Warning: File '{}' contains invalid UTF-8 data, skipping...",
+                    file_path.display()
+                );
+                None
+            } else {
+                eprintln!("Failed to read file '{}': {}", file_path.display(), err);
+                None
+            }
+        }
+    }
+}
+
+fn check_matching(file: &str, regex: &str) -> bool {
+    let re = Regex::new(&format!("(?m){}", regex)).unwrap();
+    re.is_match(file)
+}
+
+fn list_files_in_folder(folder_path: &Path) -> Result<Vec<PathBuf>, std::io::Error> {
+    let mut file_paths = Vec::new();
+
+    for entry in fs::read_dir(folder_path)? {
+        let entry = entry?;
+        let path = entry.path();
+
+        if path.is_symlink() {
+            continue;
+        }
+
+        if path.is_file() {
+            file_paths.push(path.clone());
+        } else if path.is_dir() {
+            if let Ok(sub_paths) = list_files_in_folder(&path) {
+                file_paths.extend(sub_paths);
+            }
+        }
+    }
+
+    Ok(file_paths)
+}
+
+fn all_files(folder_path: &Path) -> Vec<PathBuf> {
+    list_files_in_folder(folder_path).unwrap_or_default()
+}
+
+fn separate_regex_matching_files(
+    regex_file_path: &Path,
+    target_path: &Path,
+    include_patterns: Option<Vec<String>>,
+    exclude_patterns: Option<Vec<String>>,
+) -> Vec<PathBuf> {
+    let target_file_paths = all_files(target_path);
+    let regex = match read_file_content(regex_file_path) {
+        Some(regex_content) => regex_content,
+        None => return Vec::new(),
+    };
+
+    let include_globs = include_patterns.unwrap_or(vec!["**/*".to_string()]);
+    let exclude_globs = exclude_patterns.unwrap_or(Vec::new());
+
+    target_file_paths
+        .into_iter()
+        .filter(|target_file_name| {
+            let file_name = target_file_name.file_name().unwrap().to_str().unwrap();
+
+            let mut include_match = false;
+            for include_glob in &include_globs {
+                let pattern = glob::Pattern::new(include_glob).unwrap();
+                if pattern.matches(file_name) {
+                    include_match = true;
+                    break;
+                }
+            }
```

Comment:
This could be more stated more tersely using https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.any rather than through sentinel values like and break statements.